### DISCUSSION TITLE
PIP-16: Adding events in Pactus transactions

### DISCUSSION THREAD

**Discussion by ja7ad** [pactus-project/PIPs#73] [PIPs]
PIP-16: [Adding events in Pactus transactions](https://pips.pactus.org/PIPs/pip-16)

**Comment by b00f** [comment]
We don't have smart contract now.
I am not sure I understood it correctly, but can't we use nanomsg for future events for smart contracts?

**Reply by ja7ad** [reply]
Nope we need scalable protocol (websocket) for realtime messages.

**Reply by kehiy** [reply]
@b00f @Ja7ad 

Here is a short explanation for this (maybe I added them to PIP itself):

when there is a new state change, we need to provide a way for contracts to publish that. if we only publish it on something like nano message it's not too expensive and simple. but there are 2 main issues:

1. maybe some implementations won't utilize a nano message in the future.
2. it's the fire and forgets method and we don't keep this event/logs.

Publishing it in ways like nano message, WebSocket, and ... will have these problems.

All blockchains utilize a part of transactions and they publish contract logs there. (each protocol has some difference BTW)

In this way, everyone in every implementation has a lot of approaches to listen on them. we can publish them on nano message and WebSocket and some other implementation on any other protocols. and also they have a way to track logs/events directly on blockchain!

so we don't rely on some protocol for our logs. we publish contract state changes (which are emitted by the contract itself) on the blockchain. 

In general, we need to think more about The Pactus events.

**Reply by ja7ad** [reply]
We currently decide for the future and use a scalable protocol for communication with chan to get realtime events.

Websocket is a known protocol for web and other clients and easy for development.

I think better for the next major version we deprecated nano messages and replaced with websocket.

**Reply by b00f** [reply]
@Ja7ad, you are absolutely right about the scalable solution, and @kehiy, your point is also correct regarding nano messages (fire and forget).

However, there is one principle we must follow: **"we need to keep the core simple yet powerful"**.

So, I would like you to consider another solution: plugins. 
Each node could have some optional plugins. For example, a WebSocket plugin could listen to Nanomsg and serve clients via a WebSocket. In this case, the core remains simple, but we provide users with the option to run plugins.

**Reply by kehiy** [reply]
I know that we can use WebSocket, nano message, and ...
and it's just a protocol to serve this event and doesn't matter too much at this point.

What I'm trying to say, is that we need to have this `event/log/message/reply/call-back` entity on our core protocol to keep track of them and define a structure for them.

Or if we don't want to do that at all, so whats the reason? IMO the only reason we can have is they aren't going to be used after they are emitted and received by consumers. but if there are any scenarios in which we need them in the future, so we have to make them a core concept. then publishing them with different protocols is not so complex and hard.

**Reply by ja7ad** [reply]
@kehiy Every events have topic hash (sha256) base on name, input and output:

```none
Foobar(input1 str, input2 int) string
```

Hash

```
59100e793a81333cf2c6ff30777e441b7317ee092a4716310524e1adde265b75
```

**Reply by kehiy** [reply]
@Ja7ad Which events? That's an example from other networks or just a model as suggestion?

If you mean as an structure for our events, what do you think about adding a bytes parameter? So they can emit any structure they wish.

**Reply by ja7ad** [reply]
This model is in Ethereum to find event easy

**Reply by b00f** [reply]
Pactus uses NanoMsg for emitting events. Each event has a key and a corresponding value. 
Keys are hardcoded and defined [here](https://github.com/pactus-project/pactus/blob/be881dfa9a9377916737b3aecfd759cca510b881/www/nanomsg/event/event.go#L14). 

If we add smart contracts, we can include more events. 
NanoMsg fires events and then forgets them. 
It doesn't have any memory to keep them; it just sends events to anyone who is interested.

However, we can have an extension or service that stores NanoMsg events and then replays them. 
For example, a user may ask the event service to get all events related to smart `contract_A` after block number `N`.

**Reply by b00f** [reply]
@Ja7ad category is missed for this PIP.

**Reply by kehiy** [reply]
@b00f After checking more protocols and more research, I found that this fire-and-forget model in Pactus is good and it's enough even for contracts. I can't see a reason to keep `events`.

Our nano message can be improved and provide subscribing for specific event also. but keeping events on the chain is not necessary and an off-chain fire-and-forget model is good enough.

If you guys agree, I think we can change the status to withdrawn.


> Note: if someone searches for an on-chain event in the past, we have APIs and it's possible without events at all I think. But there are more complex situations for dApps to be considered that we can disscuess.

**Reply by b00f** [reply]
Yes, this model is good and we can implement more complicated events on top of that

**Comment by b00f** [comment]
@kehiy Thanks for drafting this PIP
This PIP is replaced by PIP-35. Status can be changed to `rejected`.
